debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java

```diff
@@ -0,0 +1,210 @@
+    private final JdbcConnection jdbcConnection;
+    private final HistorizedRelationalDatabaseSchema schema;
+
+    public HistorizedRelationalSnapshotChangeEventSource(RelationalDatabaseConnectorConfig connectorConfig, OffsetContext previousOffset, JdbcConnection jdbcConnection, HistorizedRelationalDatabaseSchema schema) {
+        this.connectorConfig = connectorConfig;
+        this.previousOffset = previousOffset;
+        this.jdbcConnection = jdbcConnection;
+        this.schema = schema;
+    }
+
+    @Override
+    public SnapshotResult execute(ChangeEventSourceContext context) throws InterruptedException {
+        // for now, just simple schema snapshotting is supported which just needs to be done once
+        if (previousOffset != null) {
+            LOGGER.debug("Found previous offset, skipping snapshotting");
+            return SnapshotResult.completed(previousOffset);
+        }
+
+        Connection connection = null;
+
+        try (SnapshotContext ctx = prepare(context)) {
+            LOGGER.info("Snapshot step 1 - Preparing");
+
+            connection = jdbcConnection.connection();
+            connection.setAutoCommit(false);
+
+            LOGGER.info("Snapshot step 2 - Determining captured tables");
+
+            // Note that there's a minor race condition here: a new table matching the filters could be created between
+            // this call and the determination of the initial snapshot position below; this seems acceptable, though
+            determineCapturedTables(ctx);
+
+            LOGGER.info("Snapshot step 3 - Locking captured tables");
+
+            if (!lockTables(context, ctx)) {
+                return SnapshotResult.aborted();
+            }
+
+            LOGGER.info("Snapshot step 4 - Determining snapshot offset");
+
+            determineSnapshotOffset(ctx);
+
+            LOGGER.info("Snapshot step 5 - Reading structure of captured tables");
+
+            if (!readTableStructure(context, ctx)) {
+                return SnapshotResult.aborted();
+            }
+
+            LOGGER.info("Snapshot step 6 - Persisting schema history");
+
+            if (!createSchemaChangeEventsForTables(context, ctx)) {
+                return SnapshotResult.aborted();
+            }
+
+            return SnapshotResult.completed(ctx.offset);
+        }
+        catch(RuntimeException e) {
+            throw e;
+        }
+        catch(Exception e) {
+            throw new RuntimeException(e);
+        }
+        finally {
+            rollbackTransaction(connection);
+
+            LOGGER.info("Snapshot step 7 - Finalizing");
+
+            complete();
+        }
+    }
+
+    /**
+     * Prepares the taking of a snapshot and returns an initial {@link SnapshotContext}.
+     */
+    protected abstract SnapshotContext prepare(ChangeEventSourceContext changeEventSourceContext) throws Exception;
+
+    private void determineCapturedTables(SnapshotContext ctx) throws Exception {
```

Comment:
@gunnarmorling SQL Server needs the list of captured tables even for streaming mode. Would it be possible to move it into database schema class?